### DISCUSSION TITLE
How to use multiple forms on a single page?

### DISCUSSION THREAD

**Discussion by dcastil** [remix-run/remix#1569] [Q&A] [answered]
Hi Remix team and community! 👋 

I'm just learning about Remix and wondering how to do loading states and optimistic updates when there are multiple forms on the same page.

The [`useTransition`](https://remix.run/docs/en/v1/api/remix#usetransition) hook doesn't seem to be tied to any specific form as indicated in [this discussion](https://github.com/remix-run/remix/discussions/1294). My understanding is that if I have multiple forms on a page (e.g. a newsletter subscribe form and a contact form), the `useTransition` hook would return the same data for both forms which makes the API somewhat awkward to work with.

Also, as far as I understand there can be just a single action defined per page, so I'm not even sure whether I could implement the serverside logic without opting out of the Remix form handling.

How do you recommend dealing with multiple forms on the same page without them stepping onto each others' feet? Or is there something to improve that on the roadmap?

**Comment by hollandThomas** [comment, marked as answer]
It gets a bit verbose but here's how you can do it:

```tsx
import type { ActionFunction } from 'remix';
import { Form, json, useTransition } from 'remix';

enum FormNames {
  FIRST_NAME_FORM = 'firstNameForm',
  LAST_NAME_FORM = 'lastNameForm',
}

export const action: ActionFunction = async ({ request }) => {
  await new Promise((res) => setTimeout(res, 1000));

  const formData = await request.formData();

  const formName = formData.get('formName');

  switch (formName) {
    case FormNames.FIRST_NAME_FORM:
      console.log('handle firstName form server side');
      return json({ firstName: formData.get('firstName') });

    case FormNames.LAST_NAME_FORM:
      console.log('handle lastName form server side');
      return json({ firstName: formData.get('lastName') });

    default:
    // handle default server side
  }
};

const Forms = () => {
  const transition = useTransition();

  const formName = transition.submission?.formData.get('formName');

  let firstNameFormText = 'Submit first name';
  let lastNameFormText = 'Submit last name';
  switch (formName) {
    case FormNames.FIRST_NAME_FORM:
      firstNameFormText =
        transition.state === 'submitting'
          ? 'Saving...'
          : transition.state === 'loading'
          ? 'Loading...'
          : 'Submit first name';

      break;
    case FormNames.LAST_NAME_FORM:
      lastNameFormText =
        transition.state === 'submitting'
          ? 'Saving...'
          : transition.state === 'loading'
          ? 'Loading...'
          : 'Submit last name';
      break;
    default:
    // handle default
  }

  return (
    <div>
      <Form method='post'>
        <label>
          First name
          <input name='firstName' type='text' />
        </label>
        <button name='formName' value={FormNames.FIRST_NAME_FORM}>
          {firstNameFormText}
        </button>
      </Form>
      <Form method='post'>
        <label>
          Last name
          <input name='lastName' type='text' />
        </label>
        <button name='formName' value={FormNames.LAST_NAME_FORM}>
          {lastNameFormText}
        </button>
      </Form>
    </div>
  );
};

export default Forms;
```

**Reply by dcastil** [reply]
Thanks for the example! I really hope that Remix can provide a better way to do this in the future. 😄

**Reply by yayakyogi** [reply]
Very helpfull, Thank you!

**Comment by na2hiro** [comment]
If multiple submission could/should happen at the same time, one `Form`s will "steal" the navigation from the other and it results in pending state rolled back, which is not a good UX.

In such cases, using `useFetcher().Form` for each type of submission would work nicely. This is the non-navigating version of `Form`. [Remix Single: Concurrent Mutations w/ useFetcher](https://www.youtube.com/watch?v=vTzNpiOk668) video describes well about it.

For references for different types of requests, [this cheatsheet](https://gist.github.com/na2hiro/136c67bbe01526eb71911d37990ed445) might help.

**Comment by ccssmnn** [comment]
You can also use hidden inputs to identify the form on the server.

```javascript
// in your action
const requestType = form.get("request-type");
switch (requestType) {
  case "change-name":
    const name = form.get("name");
    ...
  case ...
}

// in your markup
<form method="post>
  <input value="change-name" name="request-type" readOnly hidden />
  ...
</form>
```

**Comment by luispuig** [comment]
Could this proposal https://github.com/remix-run/remix/discussions/2782 a good option?

**Comment by DimaAmega** [comment]
You can try [remix-validated-form](https://www.remix-validated-form.io/reference/validated-form#subaction) form library

**Comment by karthik2265** [comment]
did remix add any improvements to handle this?

**Reply by sergiodxa** [reply]
No, its up to you how you handle them
